lld/ELF/Arch/EraVM.cpp

```diff
@@ -0,0 +1,75 @@
+//===- EraVM.cpp ----------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "Symbols.h"
+#include "Target.h"
+#include "lld/Common/ErrorHandler.h"
+#include "llvm/BinaryFormat/ELF.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/Endian.h"
+
+using namespace llvm;
+using namespace llvm::object;
+using namespace llvm::support::endian;
+using namespace llvm::ELF;
+using namespace lld;
+using namespace lld::elf;
+
+namespace {
+class EraVM final : public TargetInfo {
+public:
+  EraVM();
+  RelExpr getRelExpr(RelType type, const Symbol &s,
+                     const uint8_t *loc) const override;
+  void relocate(uint8_t *loc, const Relocation &rel,
+                uint64_t val) const override;
+};
+} // namespace
+
+EraVM::EraVM() {
+  defaultImageBase = 0;
+
+  // "panic" is 8 bytes as any other EraVM instruction, but 4 bytes required.
+  // trapInstr = {0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x04, 0x31};
+}
+
+RelExpr EraVM::getRelExpr(RelType type, const Symbol &s,
+                          const uint8_t *loc) const {
+  return R_ABS;
+}
+
+void EraVM::relocate(uint8_t *loc, const Relocation &rel, uint64_t val) const {
+  auto *definedSym = dyn_cast<Defined>(rel.sym);
+  assert(definedSym && "Defined rel.sym expected");
+  uint64_t addressInSection =
+      definedSym->getVA(rel.addend) - definedSym->section->getVA();
+
+  auto add16scaled = [loc, &rel](uint64_t value, uint64_t scale) {
+    uint64_t scaledValue = value / scale;
+    checkAlignment(loc, value, scale, rel);
+    checkIntUInt(loc, scaledValue, 16, rel);
+    write16be(loc, read16be(loc) + scaledValue);
+  };
```

Comment:
As I see the 'val' param is not used in the relocate(). To get in the context, could you please, describe how the `R_ERAVM_16_SCALE*` relocations are supposed to be resolved. If you can also remind in which cases the compiler emits these relocations, that would be great.

Reply:
> As I see the 'val' param is not used in the relocate().

In my understanding, it was easier to recalculate the value I had to use as `definedSym->getVA(rel.addend) - definedSym->section->getVA()`, but I will recheck. Anyway, this initial implementation has to be fine-tuned according to expectations of the VM, I guess.

> To get in the context, could you please, describe how the R_ERAVM_16_SCALE* relocations are supposed to be resolved. If you can also remind in which cases the compiler emits these relocations, that would be great.

At now, these are only explained in #509 (see [EraVMFixupKinds.h](https://github.com/matter-labs/era-compiler-llvm/pull/509/files#diff-729b54ebf8139b7ea94470c9ef5435be7bfbc4c3c0c748ee43d3c6d6f58257eaR23), these two fixups corespond to `R_ERAVM_16_SCALE*` relocations). These fixups are generated by [EraVMMCCodeEmiter.cpp](https://github.com/matter-labs/era-compiler-llvm/pull/509/files#diff-6ca79aa8d678c0e2fd9c139a8e4fbdf9e51c8501e516f3cf3eb4ee4fefa7156eR176).

Reply:
Thank you for the clarification. The only thing I'm worrying about is `addressInSection`: we calculate a symbol address relatively to its input section. If we are going to link at least two objects files, I suppose, we need to calculate offsets , for example, relative to the output .text segment.

Reply:
As for now we can ignore this issue, because we do not need to support linking of several object files.

Reply:
Thank you for pointing this out, removed the hardcoded re-computation of `val` and switched tests to using custom linker script at now in bf943f0341fe6b409f41714d001e99b7bb240af9.